src/MarkupInput.tsx

```diff
@@ -7,8 +12,21 @@ interface MarkupInputProps {
 
 export const MarkupInput = ({ markup, setMarkup }: MarkupInputProps) => {
   return (
-    <pre style={{ overflow: 'auto' }}>
-      <code>{markup}</code>
-    </pre>
+    <CodeMirror
+      value={markup}
+      options={{
+        mode: 'xml',
+        theme: 'material',
+        lineNumbers: true,
+        lineWrapping: true,
+        tabSize: 2
+      }}
+      onBeforeChange={(editor, data, value) => {
```

Comment:
What's the benefit of having these events on both `onBeforeChange` and `onChange`?

Reply:
It's a requirement of the controlled component, otherwise it fails. I guess it's to do with immutability? setting the state before the change and then after the change? i'm not 100% sure. In the controlled state at least, it keeps track of the users cursor position whereas in the uncontrolled component, it doesn't.

Reply:
Thanks dude, that's interesting.  I'd not seen both of these used for the same thing before